code/lib/types/src/modules/indexer.ts

```diff
@@ -63,7 +98,35 @@ export type DocsIndexEntry = BaseIndexEntry & {
   type: 'docs';
 };
 
+/**
+ * @deprecated This type is deprecated and will be replaced with {@link NewIndexEntry} in 8.0.
+ */
 export type IndexEntry = StoryIndexEntry | DocsIndexEntry;
+export interface NewBaseIndexEntry {
```

Comment:
It's annoying that we have to temporarily introduce this new type, but it's the only non-breaking way I can think of. We'll replace `IndexEntry` with `NewIndexEntry` in 8.0.

If users migrate now, they'll migrate to `NewIndexEntry`, only to migrate back to `IndexEntry` in 8.0, which must be annoying.

Reply:
I think this type isn't an `IndexEntry`. And index entry is what ends up in `index.json`, so should for instance always include an `id`.

I think this type is like an `IndexEntryInput` or something?

PS -- tell me more about `key`. For CSF files we don't need that because it's expected that the loader will produce a `CSFFile` which is keyed on `storyId`. Are we expecting something different in this case?

Reply:
> I think this type isn't an `IndexEntry`. And index entry is what ends up in `index.json`, so should for instance always include an `id`.
> 
> I think this type is like an `IndexEntryInput` or something?

I see what you mean, of course! I've changed it to that, then `IndexEntry` also won't be deprecated.

> PS -- tell me more about `key`. For CSF files we don't need that because it's expected that the loader will produce a `CSFFile` which is keyed on `storyId`. That is the loader responsible for mapping the pure ESM module exports to an object keyed on `storyId`, so we can look up the story annotations only knowing the `storyId`.
> 
> Are we expecting something different in this case?

This type is taken directly from the [Indexer RFC](https://github.com/storybookjs/storybook/discussions/23176#Implementation), which also shows example indexers that makes use of the `key` property.
I'm assuming that this is added to make indexers and story definitions more flexible, in that the export doesn't need to match the story ID, but I'd like @shilman to elaborate here.

Are you saying that if we introduce this, we need to make the annotation lookup smarter, because then it can't just rely on the `id`, it also needs to know the `key`, which I guess would mean we need the `key` in the index as well?

Reply:
>  I guess would mean we need the key in the index as well?

Yes, if the plan is to read the export named `key` in the browser when constructing the `NormalizedCsfFile`. However, I'm not sure if that makes sense because I'm pretty sure you are going to need to transform the raw module exports to get something that looks like CSF anyway (either in a loader or a runtime transformer ala [this comment](https://github.com/storybookjs/storybook/discussions/23176#discussioncomment-6267478)).

If you are already transforming the file into CSF, why wouldn't you transform the export names to mean this wasn't required (ie. so they match up with the story ids)? Perhaps I am missing a use case though.

Reply:
It's been awhile since I wrote the RFC but I think the rationale was twofold.

First, the consumer of the data needs to have the export name to be able to reference the story:

```
export default { title: 'Button' };
export const Primary = {};
```

This would generate a story with ID `button--primary`. But we don't know if the export key is `Primary`, `primary`, or something else. I'm not sure how this is handled today, but I felt that just providing the key directly would be more straightforward.

Second, it is more convenient to provide the key and have the calling code generate the story ID automatically from the export. And if you provide the ID that can take precedence.

If it's causing problems, I'm fine with using something closer to the current interface instead. But I'd like to know how it actually works currently, since it seems like it would be a bit convoluted.

Reply:
I remember @tobiasdiez having [issues with generating exports in the Vue CSF addon](https://github.com/tobiasdiez/storybook-vue-addon/pull/41), because he wanted a story named `default` (pretty sensible), but that was a reserved keyword, so he had to uppercase everything to get `Default`. In that scenario, if he was able to specify the key independent of the ID, he could prepend all exports and make them like `export const vueStoryDefault`, but still define the ID as `button--default`, so the prefix wouldn't cause "ugly" IDs.

> Second, it is more convenient to provide the key and have the calling code generate the story ID automatically from the export. And if you provide the ID that can take precedence.

I agree with this, that providing a `key` and auto-generating `name` and `id` from that seems less magical than the other way around.

I propose moving on with this, and if it turns out to be super complicated because a lot of the code assumes `id ~ key` then I'll let you know and backtrack.

Reply:
I just don't know if we should be adding an extra field to the index. 

If you want to disassociate export name and id, you can just add an id field to the annotations and call the export whatever you like. It's all generated code anyway, right?

Reply:
Ok, let me try again. Currently, the indexer must produce an `id`.

Separately, the code that loads the story computes its own `id` here (`key` is called `exportName` here):

https://github.com/storybookjs/storybook/blob/5cb57a87bfc086511b9263855ca7f1bf954cfd21/code/lib/preview-api/src/modules/store/csf/normalizeStory.ts#L55

It's convoluted because the system currently assumes that the ID generation is identical across both codepaths, and if not then it will fail.

The minimal change would be that the Indexer can return a `key`, but that the code that calls the indexer computes the `id` from the `title` and `key` (unless an `id` is also provided), but that the `key` is not stored in the index.

Reply:
That seems simple enough and uncontroversial. I'm not sure how it interacts with auto title but suspect in a good way.

Reply:
This looks like a good design. To make sure I understood it correctly, let's continue the example above: The vue indexer would handle the story `export const vueStoryDefault` by returning `id = default` and `key = vueStoryDefault`. Then the url would use the id, but the loader the key, right?

In that context, it would be nice if the indexer doesn't need to handle id/key conflicts but instead that this would be handled in a central place. For example, the svelte indexer currently keeps track of all ids its encounters and prevents id collisions (e.g. https://github.com/storybookjs/addon-svelte-csf/blob/da2523a272716884667eb6107d6a5d95ab38d706/src/parser/extract-id.ts#L13-L37). In my opinion, this adds unnecessary complications for the indexer (every indexer would need to implement a similar feature) and doesn't work if multiple indexers are used (an indexer doesn't know about the ids that another indexer creates).

Reply:
> The minimal change would be that the Indexer can return a key, but that the code that calls the indexer computes the id from the title and key (unless an id is also provided), but that the key is not stored in the index.

Now that I've implemented this and thought about it some more, this still doesn't quite make sense to me @shilman.

with this solution you're still computing the export name from the `id` (because that is all there is in the index), so the `id` must directly match the export name. so if your input a `key` that matches the export, but write a custom `id` like `my-pretty-id`, then SB will throw because it can't find the `MyPrettyId` export in the CSF.
I'm not sure what the point is of the indexer being able to specify the `id` if it must match the key anyway.

Reply:
> In that context, it would be nice if the indexer doesn't need to handle id/key conflicts but instead that this would be handled in a central place. ... In my opinion, this adds unnecessary complications for the indexer (every indexer would need to implement a similar feature) and doesn't work if multiple indexers are used (an indexer doesn't know about the ids that another indexer creates).

Agree @tobiasdiez, this makes sense to me too. Stretch goal @shilman ?